### PR TITLE
BestSalesProductSearchProvider ignores sorting argument

### LINKED ISSUE #21600: BestSalesProductSearchProvider ignores sorting argument when 'order' query param is falsy
#### Describe the bug

I am trying to display best selling products from a module. For that I am leveraging `BestSalesProductSearchProvider` with random sort order. Even if I provide sorting order, it still defaults to number of sales. This is because of the implementation, which ignores provided sorting order and sets the sorting order to number of sales if `order` query parameter is falsy:

https://github.com/PrestaShop/PrestaShop/blob/1.7.6.8/src/Adapter/BestSales/BestSalesProductSearchProvider.php#L72-L74

Why is there such restriction? If I want to display best selling products block on homepage, then I cannot image having an `order` query parameter attached.

#### Expected behavior

`BestSalesProductSearchProvider` returns products in order defined by sorting parameter.

#### Steps to Reproduce

Use the following code to get raw best selling products:
```
$searchProvider = new BestSalesProductSearchProvider(
   $this->context->getTranslator()
);

$context = new ProductSearchContext($this->context);
$query = new ProductSearchQuery();
$nProducts = 20;
$query
   ->setResultsPerPage($nProducts)
   ->setPage(1);
$query->setSortOrder(SortOrder::random()); //this gets ignored

$result = $searchProvider->runQuery(
   $context,
   $query
);
```
#### Additional information

* PrestaShop version: 1.7.6.3 & 1.7.7.0
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by ghost**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | I am trying to display best selling products from a module. For that I am leveraging BestSalesProductSearchProvider with random sort order. Even if I provide sorting order, it still defaults to number of sales.
| Type?             | improvement
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/21600
| How to test?      | Install this module (https://github.com/PrestaShop/ps_bestsellers)  and check if random work
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

This PR replace : https://github.com/PrestaShop/PrestaShop/pull/24349

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/24349)

### CODE REVIEW

File: classes/ProductSale.php [hunk 1]
```diff
@@ -97,6 +97,12 @@ public static function getBestSales($idLang, $pageNumber = 0, $nbProducts = 10,
             $orderWay = 'DESC';
         }
 
+        if ($orderWay === 'random') {
```
Comment (PierreRambaud) [inline classes/ProductSale.php]: don't execute line above if orderWay is random :thinking:

Comment (ghost) [reply classes/ProductSale.php]: Hi @PierreRambaud you ask me for this lines ?

https://github.com/PrestaShop/PrestaShop/blob/develop/classes/ProductSale.php#L85-L89

Or if it's just above

https://github.com/PrestaShop/PrestaShop/blob/develop/classes/ProductSale.php#L95-L98

i added random for Validate::isOrderWay($orderWay);

Thank you

File: classes/Validate.php [hunk 2]
```diff
@@ -761,7 +761,7 @@ public static function isZipCodeFormat($zip_code)
      */
     public static function isOrderWay($way)
     {
-        return $way === 'ASC' | $way === 'DESC' | $way === 'asc' | $way === 'desc';
+        return $way === 'ASC' | $way === 'DESC' | $way === 'asc' | $way === 'desc' | $way === 'random';
```
Comment (PierreRambaud) [inline classes/Validate.php]: I don't understand why they use bitwise :facepalm: 
You should also consider `RANDOM` or test with strtolower like that
```suggestion
        $lowerWay = strtolower($way);
        return $lowerWay === 'asc' | $lowerWay === 'desc' | $lowerWay === 'random';
```

File: classes/ProductSale.php (hunk 1 above)
Comment (PierreRambaud) [reply classes/ProductSale.php]: I mean line 85 & 91 because orderBy remove $orderBy and $orderTable

File: classes/ProductSale.php [hunk 3]
```diff
@@ -83,12 +83,12 @@ public static function getBestSales($idLang, $pageNumber = 0, $nbProducts = 10,
         $orderTable = '';
 
         $invalidOrderBy = !Validate::isOrderBy($orderBy);
```
Comment (PierreRambaud) [inline classes/ProductSale.php]: From 84 tgo 120 can be simplify:
```php
<?php

if (strtolower($orderWay) === 'random') {
    $orderWay = 'RAND()';
    $orderBy = '';
} else {
    $invalidOrderBy = !Validate::isOrderBy($orderBy);
    if ($invalidOrderBy || null === $orderBy) {
        $orderBy = 'quantity';
        $orderTable = 'ps';
    } else if (($orderBy == 'date_add' || $orderBy == 'date_upd')) {
        $orderTable = 'product_shop';
    }

    $invalidOrderWay = !Validate::isOrderWay($orderWay);
    if ($invalidOrderWay || null === $orderWay || $orderBy == 'sales') {
        $orderWay = 'DESC';
    }
}
```

File: classes/Validate.php [hunk 4]
```diff
@@ -761,7 +761,9 @@ public static function isZipCodeFormat($zip_code)
      */
     public static function isOrderWay($way)
     {
-        return $way === 'ASC' | $way === 'DESC' | $way === 'asc' | $way === 'desc';
+        $lowerWay = strtolower($way);
+
+        return $lowerWay === 'asc' | $lowerWay === 'desc' | $lowerWay === 'random';
```
Comment (matks) [inline classes/Validate.php:766]: Hello, I don't think we can do this.

If I am not wrong, `isOrderWay($way)` is used to verify that the provided `$way` can be used in an SQL query using SORT. `random` cannot be used.

In https://github.com/PrestaShop/PrestaShop/blob/develop/src/Adapter/Product/AdminProductDataProvider.php#L195

```
        $sortOrder = Validate::isOrderWay($sortOrder) ? $sortOrder : 'desc';

        $filterParams = $this->combinePersistentCatalogProductFilter(array_merge(
            $post,
            ['last_offset' => $offset, 'last_limit' => $limit, 'last_orderBy' => $orderBy, 'last_sortOrder' => $sortOrder]
        ), $avoidPersistence);
        $filterParams = AdminFilter::sanitizeFilterParameters($filterParams);
```

This block above will behave weird if we modify `Validate::isOrderWay($way)` as you suggest 😉

File: classes/ProductSale.php [hunk 5]
```diff
@@ -82,19 +82,22 @@ public static function getBestSales($idLang, $pageNumber = 0, $nbProducts = 10,
...
-        if ($invalidOrderBy || null === $orderBy) {
-            $orderBy = 'quantity';
-            $orderTable = 'ps';
-        }
-
-        if ($orderBy == 'date_add' || $orderBy == 'date_upd') {
-            $orderTable = 'product_shop';
-        }
+        if (strtolower($orderWay) === 'random') {
+            $orderWay = 'RAND()';
```
Comment (matks) [inline classes/ProductSale.php:86]: Well I did not know you could do this 😄  . It's quite powerful.

**Review by matks — COMMENTED** [review]
Hello the changes seem globally OK but I think this PR might create a bug in BO product page if I am not wrong. If someone passes `random` as sort target for product page, as there is no

```
if (strtolower($orderWay) === 'random') {
            $orderWay = 'RAND()';
```
in the product page it might generate weird sql queries.

Can you check whether product page behavior is correct?

File: classes/ProductSale.php (hunk 5 above)
Comment (ghost) [reply classes/ProductSale.php:86]: Hello @matks 

I understand what you mean but I don't see why anyone would do a random if the query can't accept it.

`random` is also introduced in : 

https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Product/Search/SortOrder.php#L202-L212

For me it's a heresy to randomize the best sellers but an issue has been opened for that.

I want to change things but I don't see how?  Do you have a way to help me?

File: classes/Validate.php (hunk 4 above)
Comment (PierreRambaud) [reply classes/Validate.php:766]: The right word should be `RAND()`

File: classes/ProductSale.php (hunk 5 above)
Comment (PierreRambaud) [reply classes/ProductSale.php:86]: @matks You didn't know the `ORDER BY RAND()`? :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream: :scream:

Comment (matks) [reply classes/ProductSale.php:86]: @PierreRambaud No, I did not know.

Do I even know SQL ?

Comment (PierreRambaud) [reply classes/ProductSale.php:86]: Heart attack

Comment (PierreRambaud) [reply classes/ProductSale.php:86]: Heart attack

File: classes/Validate.php (hunk 4 above)
Comment (ghost) [reply classes/Validate.php:766]: @PierreRambaud yes but what i do for that ?

https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Product/Search/SortOrder.php#L202-L212

It's complicated and I don't think a BC here is good ?

**Review by PierreRambaud — DISMISSED** [review]
Sorry for the delay :(

**Comment by PierreRambaud** [comment]
I rebase your PR and squash it :)

**Review by atomiix — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @okom3pom,

I followed these steps:
1. Install this PR
2. Install the module `ps_bestsellers` module
3. Transplant this module to the hook `displayHome`
4. Create some valid orders with different products
5. I used a different type of customer => I have a different Sort.

- for example, for all visitors => I have the same sort (not random)
If I refresh the page => same sort
image

- for example, for all guests => I have the same sort (not random)
If I refresh the page => same sort
image

- for example, for all customers => I have the same sort (not random)
If I refresh the page => same sort
image
image

I attached a screen record
https://watch.screencastify.com/v/lORz3WQEUBuKEDuUPcHn

Can you please check and feedback.

Thanks!

**Comment by PierreRambaud** [comment]
Hi @okom3pom ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**